ledger/metrics/src/main/scala/com/daml/telemetry/TelemetryContext.scala

```diff
@@ -0,0 +1,235 @@
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => T
+  ): T
+
+  /** Runs the computation inside an OpenTelemetry scope.
+    *
+    * This is used to set the tracing metadata in the gRPC local thread context, so it
+    * can be accessed and used by the OpenTelemetry instrumentation to create complete traces.
+    *
+    * It should be used around gRPC calls, to ensure that the tracing metadata is correctly used
+    * and transferred.
+    *
+    * @param body the computation to be run in the Telemetry scope
+    * @return the result of the computation
+    */
+  def runInOpenTelemetryScope[T](body: => T): T
+
+  /** Encode the metadata of the context in a key-value map.
+    *
+    * Typically, metadata is encoded in a map to allow transporting it across process boundaries.
+    * Originally, it has been created to carry tracing metadata across boundaries, and
+    * to create complete traces.
+    *
+    * @see [[com.daml.telemetry.Telemetry.contextFromMetadata(java.util.Map)]]
+    */
+  def encodeMetadata(): jMap[String, String]
+
+}
+
+object TelemetryContext {
+
+  /** Sets the attributes extracted from the given gRPC `MethodDescriptor`.
+    */
+  def setGrpcAttributes(
+      methodDescriptor: MethodDescriptor[_, _]
+  )(implicit telemetryContext: TelemetryContext): Unit = {
+    telemetryContext.setAttribute(SemanticAttributes.RPC_SYSTEM, "grpc")
+    telemetryContext.setAttribute(
+      SemanticAttributes.RPC_SERVICE,
+      methodDescriptor.getServiceName,
+    )
+    ()
+  }
+}
+
+/** Default implementation of TelemetryContext. Uses OpenTelemetry to generate and gather traces.
+  */
+protected class DefaultTelemetryContext(protected val span: Span) extends TelemetryContext {
+
+  def setAttribute[T](key: AttributeKey[T], value: T): TelemetryContext = {
+    span.setAttribute(key, value)
+    this
+  }
+
+  override def runFutureInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => Future[T]
+  ): Future[T] = {
+    val subSpan = createSubSpan(spanName, kind, attributes: _*)
+
+    val result = body(DefaultTelemetryContext(subSpan))
+    result.onComplete {
+      case Failure(t) =>
+        subSpan.recordException(t)
+        subSpan.end()
+      case Success(_) =>
+        subSpan.end()
+    }(DirectExecutionContext)
+    result
+  }
+
+  override def runInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => T
+  ): T = {
+    val subSpan = createSubSpan(spanName, kind, attributes: _*)
+
+    try {
+      body(DefaultTelemetryContext(subSpan))
+    } finally {
+      subSpan.end()
+    }
+  }
+
+  protected def createSubSpan(
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  ): Span = {
+    val subSpan =
+      ParticipantTracer
```

Comment:
This is a problem that I'm going to address in the next PR. Are you fine with making `tracer` a parameter?

Reply:
Let's discuss this in the next PR.